# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/gfxFT2FontList.cpp
# Commit: b07887474f51
# Full Hash: b07887474f51c5d5e01cb0188c18c8618c77be57
# Author: Doug Thayer <dothayer@mozilla.com>
# Date: 2020-07-02 03:58:28
# Regressor Bug: 1627075
# File Overlap Count: 1
# Description:
#   Bug 1627075 - Route Omnijar requests through StartupCache r=froydnj
#   
#   This should be a relatively straightforward patch. Essentially, we implement
#   a wrapper class (and friends) around nsZipArchive (and friends), which transparently
#   caches entries from the underlying zip archive in the StartupCache. This will break
# ==============================================================================

diff -r 65c0e6790a33 -r b07887474f51 gfx/thebes/gfxFT2FontList.cpp
--- a/gfx/thebes/gfxFT2FontList.cpp	Wed Jul 01 17:55:38 2020 +0000
+++ b/gfx/thebes/gfxFT2FontList.cpp	Wed Jul 01 17:09:53 2020 +0000
@@ -95,14 +95,14 @@
   // here would be memory allocation, in which case mFace remains null.
   RefPtr<SharedFTFace> face;
   if (mFilename[0] != '/') {
-    RefPtr<nsZipArchive> reader = Omnijar::GetReader(Omnijar::Type::GRE);
+    RefPtr<CacheAwareZipReader> reader = Omnijar::GetReader(Omnijar::Type::GRE);
     nsZipItem* item = reader->GetItem(mFilename.get());
     NS_ASSERTION(item, "failed to find zip entry");
 
     uint32_t bufSize = item->RealSize();
     uint8_t* fontDataBuf = static_cast<uint8_t*>(malloc(bufSize));
     if (fontDataBuf) {
-      nsZipCursor cursor(item, reader, fontDataBuf, bufSize);
+      CacheAwareZipCursor cursor(item, reader, fontDataBuf, bufSize);
       cursor.Copy(&bufSize);
       NS_ASSERTION(bufSize == item->RealSize(), "error reading bundled font");
       RefPtr<FTUserFontData> ufd = new FTUserFontData(fontDataBuf, bufSize);
@@ -458,7 +458,8 @@
     } else {
       // A relative path means an omnijar resource, which we may need to
       // decompress to a temporary buffer.
-      RefPtr<nsZipArchive> reader = Omnijar::GetReader(Omnijar::Type::GRE);
+      RefPtr<CacheAwareZipReader> reader =
+          Omnijar::GetReader(Omnijar::Type::GRE);
       nsZipItem* item = reader->GetItem(mFilename.get());
       MOZ_ASSERT(item, "failed to find zip entry");
       if (item) {
@@ -469,7 +470,7 @@
         uint32_t length = item->RealSize();
         uint8_t* buffer = static_cast<uint8_t*>(malloc(length));
         if (buffer) {
-          nsZipCursor cursor(item, reader, buffer, length);
+          CacheAwareZipCursor cursor(item, reader, buffer, length);
           cursor.Copy(&length);
           MOZ_ASSERT(length == item->RealSize(), "error reading font");
           if (length == item->RealSize()) {
@@ -1159,7 +1160,7 @@
   static const char* sJarSearchPaths[] = {
       "res/fonts/*.ttf$",
   };
-  RefPtr<nsZipArchive> reader = Omnijar::GetReader(Omnijar::Type::GRE);
+  RefPtr<CacheAwareZipReader> reader = Omnijar::GetReader(Omnijar::Type::GRE);
   for (unsigned i = 0; i < ArrayLength(sJarSearchPaths); i++) {
     nsZipFind* find;
     if (NS_SUCCEEDED(reader->FindInit(sJarSearchPaths[i], &find))) {
@@ -1274,7 +1275,7 @@
   }
 }
 
-void gfxFT2FontList::AppendFacesFromOmnijarEntry(nsZipArchive* aArchive,
+void gfxFT2FontList::AppendFacesFromOmnijarEntry(CacheAwareZipReader* aArchive,
                                                  const nsCString& aEntryName,
                                                  FontNameCache* aCache,
                                                  bool aJarChanged) {
@@ -1313,7 +1314,7 @@
     return;
   }
 
-  nsZipCursor cursor(item, aArchive, (uint8_t*)buffer, bufSize);
+  CacheAwareZipCursor cursor(item, aArchive, (uint8_t*)buffer, bufSize);
   uint8_t* data = cursor.Copy(&bufSize);
   MOZ_ASSERT(data && bufSize == item->RealSize(), "error reading bundled font");
   if (!data) {